src/components/molecules/Visualizer/Layers/store.ts

```diff
@@ -68,6 +73,70 @@ export class LayerStore {
     return l;
   }
 
+  #databaseLayers: Layer | undefined;
+  #appendedLayersData: AppendedLayerData[];
+
+  #mapLayers = () => {
+    this.#flattenLayers = flattenLayers(this.#root?.children ?? []);
+    this.#map = new Map(this.#flattenLayers.map(l => [l.id, l]));
+    this.#pmap = new Map(this.#flattenLayers.map(l => [l.id, this.#pluginLayer(l)]));
+  };
+
+  #updateLayers = () => {
+    this.#root = this.#databaseLayers ?? { id: "", children: [] };
+    this.#proot = this.#pluginLayer(this.#root);
+    this.#mapLayers();
+    this.#appendedLayersData.map(l => this.#insertAppendedLayer(l));
+  };
+
+  #appendedResIds: string[] = [];
+  #getUniqueAppendedResId = () => {
+    const genResId = () =>
+      "_xxxxxxxxxxxxxxxxxxxxxxxxx".replace(/[x]/g, function () {
+        return ((Math.random() * 16) | 0).toString(16);
+      });
+    let uniqueAppendedResId;
+    do {
+      uniqueAppendedResId = genResId();
+    } while (this.#appendedResIds.indexOf(uniqueAppendedResId) !== -1);
+    this.#appendedResIds.push(uniqueAppendedResId);
+    return uniqueAppendedResId;
+  };
+
+  #insertAppendedLayer = (appendedLayerData: AppendedLayerData) => {
+    (appendedLayerData.parentId
+      ? this.#map.get(appendedLayerData.parentId)
+      : this.#root
+    )?.children?.push(appendedLayerData.layer);
+    Object.setPrototypeOf(this.#proot, { children: this.#root.children });
+    this.#mapLayers();
+  };
+
+  append = (layer: Omit<Layer, "id">, parentId?: string) => {
+    const layerId = this.#getUniqueAppendedResId();
+    const appendedLayerData = {
+      layer: {
+        ...layer,
+        id: layerId,
+        pluginId: "reearth",
+      },
+      parentId,
+    };
+    if (layer.infobox?.blocks) {
+      layer.infobox.blocks.map(b => {
+        b.id = this.#getUniqueAppendedResId();
+      });
+    }
+    this.#appendedLayersData.push(appendedLayerData);
+    this.#insertAppendedLayer(appendedLayerData);
```

Comment:
Not including `this.#appendedLayersData.push(appendedLayerData);` in the `#insertAppendedLayer ` function seems a little hard to read in the code, I think